### PR TITLE
[JSC] Add ProxyObjectLoad IC

### PR DESCRIPTION
**Pull request by Constellation**
#### a328f5aa59d892239b1bb119de0bfeb7169cc4f8
<pre>
[JSC] Add ProxyObjectLoad IC
<a href="https://bugs.webkit.org/show_bug.cgi?id=244362">https://bugs.webkit.org/show_bug.cgi?id=244362</a>

Reviewed by Justin Michaud.

This patch adds optimization for Proxy&apos;s [[Get]].

1. This patch adds ProxyObjectLoad IC, which detects ProxyObject and calls @performProxyObjectGet JS function. This handles
   property lookup in handler / target significantly faster. And we can jump to ProxyObject [[Get]] path instead of going to
   C++ and going back to JS.
2. This patch adds GetByValWithThis IC, it means we thisGPR register handling is added to GetByVal IC too. And we start using
   it in Baseline, DFG, and FTL. This IC is added since @performProxyObjectGet needs this operation in a super faster way.

Still, we have a lot of good optimization: DFG layer conversion from GetByValWithThis to GetByIdWithThis. Handling ProxyObjectLoad
in DFG and inline @performProxyObjectGet. But for now, let&apos;s just do optimization via IC since it can make baseline speed faster
in all JIT tiers.

This improves Proxy [[Get]] significantly in microbenchmarks. And we observed 8% improvement in JetStream2 chai-wtb since it heavily uses Proxy.

                                       ToT                     Patched

    proxy-get                   608.4166+-0.6237     ^    147.8951+-1.1855        ^ definitely 4.1138x faster
    proxy-get-miss-handler      304.8690+-0.8261     ^     48.4809+-0.1115        ^ definitely 6.2884x faster

* JSTests/microbenchmarks/proxy-get-miss-handler.js: Added.
* JSTests/microbenchmarks/proxy-get.js: Added.
(get target):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/ProxyHelpers.js: Added.
(linkTimeConstant.performProxyObjectGet):
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::needsScratchFPR const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::generateWithGuard):
(JSC::AccessCase::generateImpl):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
(JSC::SharedJITStubSet::Hash::Key::Key):
(JSC::SharedJITStubSet::Hash::Key::operator==):
(JSC::SharedJITStubSet::Searcher::Translator::equal):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/GetterSetterAccessCase.cpp:
(JSC::GetterSetterAccessCase::emitDOMJITGetter):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecode/ModuleNamespaceAccessCase.cpp:
(JSC::ModuleNamespaceAccessCase::emit):
* Source/JavaScriptCore/bytecode/PolymorphicAccess.cpp:
(JSC::AccessGenerationState::makeDefaultScratchAllocator):
(JSC::PolymorphicAccess::regenerate):
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/PolymorphicAccess.h:
(JSC::AccessGenerationState::thisGPR const): Deleted.
(JSC::AccessGenerationState::prototypeGPR const): Deleted.
(JSC::AccessGenerationState::propertyGPR const): Deleted.
* Source/JavaScriptCore/bytecode/ProxyObjectAccessCase.cpp: Added.
(JSC::ProxyObjectAccessCase::ProxyObjectAccessCase):
(JSC::ProxyObjectAccessCase::create):
(JSC::ProxyObjectAccessCase::clone const):
(JSC::ProxyObjectAccessCase::emit):
(JSC::ProxyObjectAccessCase::dumpImpl const):
* Source/JavaScriptCore/bytecode/ProxyObjectAccessCase.h: Added.
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::appropriateOptimizingGetByFunction):
(JSC::appropriateGetByFunction):
(JSC::tryCacheGetBy):
(JSC::resetGetBy):
* Source/JavaScriptCore/bytecode/Repatch.h:
* Source/JavaScriptCore/bytecode/StructureStubInfo.cpp:
(JSC::StructureStubInfo::reset):
(JSC::slowOperationFromUnlinkedStructureStubInfo):
(JSC::StructureStubInfo::initializeFromUnlinkedStructureStubInfo):
(JSC::StructureStubInfo::initializeFromDFGUnlinkedStructureStubInfo):
(JSC::StructureStubInfo::checkConsistency):
* Source/JavaScriptCore/bytecode/StructureStubInfo.h:
(JSC::StructureStubInfo::thisValueIsInExtraGPR const):
(JSC::StructureStubInfo::prototypeGPR const):
(JSC::StructureStubInfo::propertyGPR const):
(JSC::StructureStubInfo::propertyTagGPR const):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::BytecodeIntrinsicNode::emit_intrinsic_getByValWithThis):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGJITCode.h:
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::link):
* Source/JavaScriptCore/dfg/DFGJITCompiler.h:
(JSC::DFG::JITCompiler::addGetByValWithThis):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByValWithThis):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByValWithThis):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileGetByValWithThis):
* Source/JavaScriptCore/jit/BaselineJITRegisters.h:
* Source/JavaScriptCore/jit/ICStats.h:
* Source/JavaScriptCore/jit/IntrinsicEmitter.cpp:
(JSC::IntrinsicGetterAccessCase::emitIntrinsicGetter):
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::privateCompileMainPass):
(JSC::JIT::privateCompileSlowCases):
(JSC::JIT::link):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITInlineCacheGenerator.cpp:
(JSC::JITGetByValGenerator::finalize):
(JSC::JITGetByValWithThisGenerator::JITGetByValWithThisGenerator):
(JSC::JITGetByValWithThisGenerator::generateFastPath):
(JSC::JITGetByValWithThisGenerator::generateEmptyPath):
(JSC::JITGetByValWithThisGenerator::finalize):
* Source/JavaScriptCore/jit/JITInlineCacheGenerator.h:
(JSC::JITInlineCacheGenerator::JITInlineCacheGenerator): Deleted.
(JSC::JITByIdGenerator::JITByIdGenerator): Deleted.
(JSC::JITInByValGenerator::JITInByValGenerator): Deleted.
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::getByValWithThis):
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/jit/JITOperations.h:
* Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
(JSC::JIT::emit_op_get_by_val_with_this):
(JSC::JIT::emitSlow_op_get_by_val_with_this):
(JSC::JIT::slow_op_get_by_val_with_this_callSlowOperationThenCheckExceptionGenerator):
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/Operations.h:
(JSC::getByValWithIndexAndThis):
* Source/JavaScriptCore/runtime/ProxyObject.h:

Canonical link: <a href="https://commits.webkit.org/254092@main">https://commits.webkit.org/254092@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0ecaa4bf767069e12befcd2a73d60057fb4b0dda

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87997 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32158 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97144 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152625 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30527 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26474 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80093 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91888 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93608 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24586 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/74659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24563 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/79526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79754 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/28166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73505 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/28263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14507 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26119 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2868 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/31287 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/74659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76347 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/30236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33765 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16942 "Passed tests") |

### CODE REVIEW

**Comment by MenloDorian** [comment]
> #### [704fb05](https://github.com/WebKit/WebKit/commit/704fb05ab63552d1b0429b2a4977bf29fa268c1b)
\> 1. This patch adds ProxyObjectLoad IC, which detects ProxyObject and calls @performProxyObjectGet JS function. This handles
>    property lookup in handler / target significnatly faster. And we can jump to ProxyObject [[Get]] path instead of going to

/significnatly/significantly/

> 2. This patch adds GetByValWithThis IC, it means we thisGPR register handling is added to GetByVal IC too. And we start using

/means we thisGPR register/means thisGPR register/ ?


> Still, we have a lot of good optimization: DFG layer conversion from GetByValWithThis to GetByIdWithThis. Handling ProxyObjectLoad

/Still, we have a lot of good optimization/We still have a lot of opportunity for optimization/

File: Source/JavaScriptCore/bytecode/Repatch.cpp [hunk 1]
```diff
@@ -386,6 +391,17 @@ static InlineCacheAction tryCacheGetBy(JSGlobalObject* globalObject, CodeBlock*
             if (auto moduleNamespaceSlot = slot.moduleNamespaceSlot())
                 newCase = ModuleNamespaceAccessCase::create(vm, codeBlock, propertyName, jsCast<JSModuleNamespaceObject*>(baseCell), moduleNamespaceSlot->environment, ScopeOffset(moduleNamespaceSlot->scopeOffset));
         }
+
+        if ((!propertyName.uid()->isSymbol() || static_cast<SymbolImpl&>(*propertyName.uid()).isPrivate()) && baseCell->inherits<ProxyObject>()) {
```
Comment (justinmichaud) [inline Source/JavaScriptCore/bytecode/Repatch.cpp]: Why can't we cache symbols?

Comment (justinmichaud) [reply Source/JavaScriptCore/bytecode/Repatch.cpp]: Non-private symbols are symbols created by the user, right?

File: Source/JavaScriptCore/bytecode/StructureStubInfo.cpp [hunk 2]
```diff
@@ -683,10 +702,15 @@ void StructureStubInfo::initializeFromDFGUnlinkedStructureStubInfo(const DFG::Un
 #if ASSERT_ENABLED
 void StructureStubInfo::checkConsistency()
 {
-    if (thisValueIsInExtraGPR()) {
+    switch (accessType) {
```
Comment (justinmichaud) [inline Source/JavaScriptCore/bytecode/StructureStubInfo.cpp:705]: I wish we had a better story for specifying the state machine of used registers

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp [hunk 3]
```diff
@@ -1186,6 +1186,10 @@ class FixupPhase : public Phase {
             break;
         }
 
+        case GetByValWithThis: {
+            break;
```
Comment (justinmichaud) [inline Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:1190]: Should we predict something here?

File: Source/JavaScriptCore/runtime/Operations.h [hunk 4]
```diff
@@ -895,15 +895,29 @@ ALWAYS_INLINE JSValue jsBitwiseXor(JSGlobalObject* globalObject, JSValue v1, JSV
 ALWAYS_INLINE EncodedJSValue getByValWithIndex(JSGlobalObject* globalObject, JSCell* base, uint32_t index)
 {
     if (base->isObject()) {
-        JSObject* object = asObject(base);
-        if (JSValue result = object->tryGetIndexQuickly(index))
+        if (JSValue result = asObject(base)->tryGetIndexQuickly(index))
```
Comment (justinmichaud) [inline Source/JavaScriptCore/runtime/Operations.h]: Could this just call getByValWithIndexAndThis with this = base?

File: Source/JavaScriptCore/runtime/JSGlobalObject.cpp [hunk 5]
```diff
@@ -365,6 +366,36 @@ JSC_DEFINE_HOST_FUNCTION(hasOwnLengthProperty, (JSGlobalObject* globalObject, Ca
...
+            RETURN_IF_EXCEPTION(scope, { });
+            if (!isSame)
+                return throwVMTypeError(globalObject, scope, "Proxy handler's 'get' result of a non-configurable and non-writable property should be the same value as the target's property"_s);
+        } else if (descriptor.isAccessorDescriptor() && !descriptor.configurable() && descriptor.getter().isUndefined()) {
+            if (!trapResult.isUndefined())
+                return throwVMTypeError(globalObject, scope, "Proxy handler's 'get' result of a non-configurable accessor property without a getter should be undefined"_s);
+        }
+    }
+
+    RELEASE_AND_RETURN(scope, JSValue::encode(jsUndefined()));
```
Comment (justinmichaud) [inline Source/JavaScriptCore/runtime/JSGlobalObject.cpp:396]: So this always returns undefined?

File: Source/JavaScriptCore/jit/JITPropertyAccess.cpp [hunk 6]
```diff
@@ -2153,6 +2153,113 @@ void JIT::emit_op_put_internal_field(const JSInstruction* currentInstruction)
 
 #if USE(JSVALUE64)
 
+void JIT::emit_op_get_by_val_with_this(const JSInstruction* currentInstruction)
```
Comment (justinmichaud) [inline Source/JavaScriptCore/jit/JITPropertyAccess.cpp:2156]: so op_get_by_val_with_this is exactly the same as op_get_by_val, but with (this != base)?

**Review by justinmichaud — APPROVED** [review]
r=me with questions. I am curious: what stops us from calling GetByIdWithThis in place of GetById in most of these places? Could we just make base == this a special case?

File: Source/JavaScriptCore/bytecode/Repatch.cpp (hunk 1 above)
Comment (Constellation) [reply Source/JavaScriptCore/bytecode/Repatch.cpp]: Oops! Right. Fixed.

Comment (Constellation) [reply Source/JavaScriptCore/bytecode/Repatch.cpp]: Oops, right. Fixed.

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp (hunk 3 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:1190]: In the future, yes. Currently, we are not doing any optimizations with speculations.

File: Source/JavaScriptCore/jit/JITPropertyAccess.cpp (hunk 6 above)
Comment (Constellation) [reply Source/JavaScriptCore/jit/JITPropertyAccess.cpp:2156]: yes, `thisValue` is different. It affects on the behavior when we hit getter.

File: Source/JavaScriptCore/runtime/JSGlobalObject.cpp (hunk 5 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/JSGlobalObject.cpp:396]: Yes, returning undefined, or throwing an error. The returned value is not used.

File: Source/JavaScriptCore/runtime/Operations.h (hunk 4 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/Operations.h]: Yeah, changed.

**Review by Constellation — COMMENTED** [review]
Thank you!

**Comment by Constellation** [comment]
@MenloDorian Fixed.

**Pull request merged**